One-off trigger + effects

rish987 opened the pull request:
I added a `PEN_UP` internal trigger that fires after the pen is lifted off of the screen. This enables "one-off" effects, for example the `one-off-eraser` effect that activates the eraser and automatically switches back to the writing pen after you're done erasing (don't need to click again).

rish987 commented on sources/effects.c:
```diff
@@ -242,6 +244,27 @@ void toggle_tool_select(int fd_touch) {
+}
+
+void pen_up(int fd_touch, int fd_pen) {
+  if (oneOffEraseSelect) {
+    deactivate_tool_eraser_select(fd_touch);
+    oneOffEraseSelect = false;
+  }
```
I haven't tested the `one-off-eraser-selection` effect on Remarkable version 3.8, so I'm not sure if this will work for you -- on version 3.9 it doesn't work at all, I have to add a delay (see [here](https://github.com/rish987/RemarkableLamyEraser/blob/f51b51cc6d9070a874b369ae9f6fbbeafafa9ea7/sources/effects.c#L293)).

rish987 replied:
Okay, having recently downgraded to 3.8 I can confirm that you do need a delay. I'm AFC for the next few days but [here is my personal development branch](https://github.com/rish987/RemarkableLamyEraser/tree/hl-toggle-long-click) if you're curious.

slotThe commented on headers/effects_data.h:
```diff
@@ -27,6 +27,17 @@ enum effects {
   NUM_EFFECTS,         // Total number of effects
 };
 
+// extra effects that do not exactly correspond to a screen position
```
I think it'd still be cleaner to put these into the `effects` enum

slotThe replied:
I think it'd still be cleaner to put these into the `effects` enum

rish987 committed:
one-off trigger for eraser, eraser selection, and highlighter

rish987 committed:
refactor trigger indexing

rish987 committed:
allow for releasing button after pen touches screen with press&hold (though does not mesh well with xochitl interface)

rish987 commented:
Hi! I've updated the PR, and made a few more changes:

1. Added the config flag `assume-toolbar-open` (see README).
2. Made the indexing more compact for triggers with count (click, long-click, press&hold) and toggle/temporary effects (using `enum effect_type`).
3. Wrote code to make it possible to let go of the button after the pen touches the screen with press&hold effects. Unfortunately, though, I think this may hit the limit of what is possible with this tool, as it doesn't work with the eraser (and the `action_tool_eraser()` hack to override `TOOL_PEN` events doesn't seem to help). It works as expected for other effects, though not reliably due to UI issues (as with other triggers). But it doesn't hurt to have the functionality implemented anyways.

Overall, I have found that the UI responsiveness has degraded a lot since upgrading to 3.9, and sometimes it just locks up and I have to open/close the toolbar, or in the worst case turn my Remarkable on/off. I've started to port some of the code here for use with [rm-hacks](https://github.com/mb1986/rm-hacks) ([see here](https://github.com/rish987/remarkable-stylus)) so that we can hopefully get around this (I know you're not the biggest fan of binary patching hahaha, but in any case this tool has been a very good "staging area" for me to test some of the things I want to see implemented there).

slotThe commented:
Thanks!
